src/track/serato/tags.cpp

```diff
@@ -200,70 +191,90 @@ QList<CueInfo> SeratoTags::getCueInfos() const {
+        for (int i = 0; i < kNumCuesInMarkersTag; i++) {
+            unsetCuesInMarkersTag.insert(i);
         }
 
-        // Take a pre-existing CueInfo object that was read from
-        // "SeratoMarkers2" from the CueMap (or a default constructed CueInfo
-        // object if none exists) and use it as template for the new CueInfo
-        // object. Then overwrite all object values that are present in the
-        // "SeratoMarkers_"tag.
-        CueInfo newCueInfo = cueMap.value(*index);
-        newCueInfo.setType(cueInfo.getType());
-        newCueInfo.setStartPositionMillis(cueInfo.getStartPositionMillis());
-        newCueInfo.setEndPositionMillis(cueInfo.getEndPositionMillis());
-        newCueInfo.setHotCueIndex(index);
-        newCueInfo.setFlags(cueInfo.flags());
-        newCueInfo.setColor(cueInfo.getColor());
-        cueMap.insert(*index, newCueInfo);
+        for (const CueInfo& cueInfo : cuesMarkers) {
+            if (!isCueInfoValid(cueInfo)) {
+                continue;
+            }
+            // Take a pre-existing CueInfo object that was read from
+            // "SeratoMarkers2" from the CueMap (or a default constructed CueInfo
+            // object if none exists) and use it as template for the new CueInfo
+            // object. Then overwrite all object values that are present in the
+            // "SeratoMarkers_"tag.
+            CueInfo& newCueInfo = cueInfo.getType() == CueType::Loop
+                    ? loopCueMap[*cueInfo.getHotCueIndex()]
+                    : hotCueMap[*cueInfo.getHotCueIndex()];
+            newCueInfo.setType(cueInfo.getType());
+            newCueInfo.setStartPositionMillis(cueInfo.getStartPositionMillis());
+            newCueInfo.setEndPositionMillis(cueInfo.getEndPositionMillis());
+            newCueInfo.setHotCueIndex(*cueInfo.getHotCueIndex());
+            newCueInfo.setFlags(cueInfo.flags());
+            newCueInfo.setColor(cueInfo.getColor());
+
+            // This cue is set in the "Serato Markers_" tag, so remove it from the
+            // set of unset cues
+            unsetCuesInMarkersTag.remove(*cueInfo.getHotCueIndex());
+        };
+
+        // Now that we know which cues should be present in the "Serato Markers_"
+        // tag but aren't, remove them from the set.
+        for (const int index : unsetCuesInMarkersTag) {
+            hotCueMap.remove(index);
+        }
+    }
 
-        // This cue is set in the "Serato Markers_" tag, so remove it from the
-        // set of unset cues
-        unsetCuesInMarkersTag.remove(*index);
-    };
+    // Serato has a separate indexing for loop and hot cues.
+    // We sort the Loops Cues into the HotCue map if the given index is free,
+    // add an offset of 8 otherwise or loop until we find a free index.
+    // Cues above index 8 are not accessible in Mixxx, only visible in waveforms
+    // During export the Mixxx indexes are kept to allow a perfect round-trip.
```

Comment:
This sounds like a regression to me. Now the ordering of hotcues and loops is quite unpredictable to the regular user.
The main target audience of this feature is people switching from Serato, where these banks are separate. In Mixxx, they are suddenly mixed.

In Serato, you have the following cues:

1. Hotcue A
2. Hotcue B
3.
4. Hotcue D
5. ...

1. Loop A
2. Loop B
3. Loop C
4. Loop D
5. ...

When importing into Mixxx, this will now suddenly become:

1. Hotcue A
2. Hotcue B
3. Loop C
4. Hotcue D
5. ...
8. Loop A
9. Loop B
10. *unassigned*
11. Loop D
12. ...

This is IMHO unexpected and should be reverted. The fixed offset of 8. makes more sense: If you have more than 8 cues and loops total, you need access to cues with index > 8 anyway.

Also, users that previously used Serato are likely to have a Serato-Style DJ controller, which features separate Hotcue and Loop banks. For example, the Roland DJ-505 mapping uses cues 1-8 for hotcues and 9-16 for loop cues. This way, these users can continue using their controller and tracks as expected.

Reply:
oof, bad timing... this was just merged :(

Reply:
I'm sorry for the bad timing

> This is IMHO unexpected and should be reverted. The fixed offset of 8. makes more sense: If you have more than 8 cues and loops total, you need access to cues with index > 8 anyway.

I fully agree, but how would you keep the offset index from accumulating then? How would tell loop cues interspersed with hotcues by mixxx and loopcues from serato apart?

Reply:
I think the *export* is still marked as *experimental*, so if we have to make amends, then it should be in the export logic, not the import logic. It's a format for a different software, so some downside is to be expected.

Serato does not have interspersed hotcues/loop cues, so the order will have to change on export anyway.
One simple way would be to restore the previous import behavior and check if all loop cues have an index > 8. If so, subtract 8 from the index during export.
This would at least allow to preserve empty loop cue slots.

Example:

Cues:

1. Hotcue A
2. Hotcue B
3. 
4. Hotcue D
5. ...

Loops:

1. Loop A
2. 
3. Loop C
4. Loop D
5. ...

Becomes this in Mixxx:

1. Hotcue A
2. Hotcue B
3. 
4. Hotcue D
5. ...
6. ...
7. ...
8. ...
9. Loop A
10.
11. Loop C
12. Loop D
13. ...

And on export the original is restored:

Cues:

1. Hotcue A
2. Hotcue B
3. 
4. Hotcue D
5. ...

Loops:

1. Loop A
2. 
3. Loop C
4. Loop D
5. ...

Reply:
Yes, this will be suboptimal for people who want to move away from Mixxx to Serato, but if we have to choose which kind of user we want to accomodate, it should be the user group that wants to migrate from Serato *to* Mixxx, instead of the group that wants to leave Mixxx anyway.

Reply:
It looks like this are two different demands. The issue in Mixxx is that we have only 8 buttons in the GUI. So from your example only Loop C is usable for all users. For me the current solution is a good compromise that at least fixes the round-trip starting from Mixxx. I am afraid a full solution requires to store the Mixxx index and the Serato index separate. This is out of scope of this PR.  Do you have an alternative solution? 
Another open issue is that the white loop cues are becoming orange after the round-trip. Do you have an idea how to fix that? 

A regular Mixxx user has probably 

1. Hotcue A
2. Hotcue B
3. ...
4. Hotcue D
5. Loop A
6. Loop B
7. 
8. Loop D

This is exported now and before as 

Hotcues
1. Hotcue A
2. Hotcue B
3. ...
4. Hotcue D

Loops
1. Loop A
2. Loop B
3. ... 
4. Loop D

After reading it back from the file it becomes before that PR: 

1. Hotcue A
2. Hotcue B
3. ...
4. Hotcue D
5. ...
6. ...
7. ...
8. ...
9. Loop A
10. Loop B
11. 
12. Loop D

The Loops are no longer accessible. And after another round-trip it is: 

1. Hotcue A
2. Hotcue B
3. ...
4. Hotcue D
5. ...
6. ...
7. ...
8. ...
9. ...
10. ...
11. ...
12. ...
13. ...
14. ...
15. ..
16. ...
17. Loop A
18. Loop B
19. 
20. Loop D

Reply:
> For me the current solution is a good compromise that at least fixes the round-trip starting from Mixxx.

Again, I'd argue that our focus should be on the people who want to switch from Serato to Mixxx, not the people who want to switch from Mixxx to Serato. The primary purpose of the *experimental* export feature is to avoid lock-in. Due to differences between Mixxx and Serato, this format is not suited for lossless transfor of cues between two Mixxx instances.

For Mixxx-only users, the proper solution would be to devise a custom format, e.g. something bson-based maybe.

> Another open issue is that the white loop cues are becoming orange after the round-trip.

Loop cues don't have colors in Serato, they are all blue. I hypothesized that there is a color field and just the GUI support is missing in Serato, because there is a field that would indicate the color blue, but setting a different "color" in the metadata did not work. We should probably remove color support for loop cues altogether, because it is not supported.